src/miner_gateway_port.erl

```diff
@@ -0,0 +1,148 @@
+%%%-------------------------------------------------------------------
+%% @doc miner gateway service manager
+%% Start and monitor the port that runs the external rust-based gateway
+%% @end
+%%%-------------------------------------------------------------------
+-module(miner_gateway_port).
+
+-behaviour(gen_server).
+
+-export([
+    start_link/1,
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2
+]).
+
+-record(state, {
+    keypair,
+    transport=tcp,
+    host="localhost",
+    port=4468,
+    monitor,
+    os_pid,
+    tcp_port
+}).
+
+-define(CONNECT_RETRY_WAIT, 100).
+
+start_link(Options) when is_list(Options) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Options], []).
+
+init([Options]) ->
+    Keypair = proplists:get_value(keypair, Options),
+    TcpPort = proplists:get_value(port, Options, 4468),
+    Host = proplists:get_value(host, Options, "localhost"),
+    Transport = proplists:get_value(transport, Options, tcp),
+
+    process_flag(trap_exit, true),
+
+    State = open_gateway_port(Keypair, Transport, Host, TcpPort),
+    {ok, State}.
+
+handle_call(_Msg, _From, State) ->
+    lager:debug("unhandled call ~p", [_Msg]),
+    {noreply, State}.
+
+handle_cast(_Msg, State) ->
+    lager:debug("unhandled cast ~p", [_Msg]),
+    {noreply, State}.
+
+handle_info({Port, {exit_status, Status}}, #state{port = Port} = State) ->
+    lager:warning("gateway-rs process ~p exited with status ~p, restarting", [Port, Status]),
+    ok = cleanup_port(State),
+    NewState = open_gateway_port(State#state.keypair, State#state.transport, State#state.host, State#state.tcp_port),
+    ok = miner_gateway_ecc_worker:reconnect(),
+    {noreply, NewState};
+handle_info({Port, {data, LogMsg}}, #state{port = Port} = State) ->
+    Lines = binary:split(LogMsg, <<"\n">>, [global]),
```

Comment:
If the last line is incomplete doesn't this mean you may log a partial line?

Reply:
testing this out i didn't see any weird behavior if the final complete log message _didn't_ include a trailing `\n` so the final item in the split list should be as complete or incomplete as whatever the gateway sends over stdout; unless i stuffed any incomplete lines into the state and waited to see if they were completed how would i ensure i incomplete lines weren't logged? or are you thinking it'd be better to _not_ to log incomplete lines at all?